src/Microsoft.AspNet.Hosting/Program.cs

```diff
@@ -34,34 +34,19 @@ public void Main(string[] args)
             config.AddEnvironmentVariables();
             config.AddCommandLine(args);
 
-            var host = new WebHostBuilder(_serviceProvider, config).Build();
-            var serverShutdown = host.Start();
-            var loggerFactory = host.ApplicationServices.GetRequiredService<ILoggerFactory>();
-            var appShutdownService = host.ApplicationServices.GetRequiredService<IApplicationShutdown>();
-            var shutdownHandle = new ManualResetEvent(false);
-
-            appShutdownService.ShutdownRequested.Register(() =>
+            var host = new WebHostBuilder(_serviceProvider, config)
+                .UseServices(x => x.AddSingleton<IHostingKeepAlive, ConsoleHostingKeepAlive>())
+                .Build();
+            using (host.Start())
             {
-                try
-                {
-                    serverShutdown.Dispose();
-                }
-                catch (Exception ex)
-                {
-                    var logger = loggerFactory.CreateLogger<Program>();
-                    logger.LogError("Dispose threw an exception.", ex);
-                }
-                shutdownHandle.Set();
-            });
+                var appShutdownService = host.ApplicationServices.GetRequiredService<IApplicationShutdown>();
+                var hostingKeepAlives = host.ApplicationServices.GetRequiredServices<IHostingKeepAlive>();
 
-            var ignored = Task.Run(() =>
-            {
-                Console.WriteLine("Started");
-                Console.ReadLine();
-                appShutdownService.RequestShutdown();
-            });
+                foreach (var hostingKeepAlive in hostingKeepAlives)
+                    hostingKeepAlive.Setup(appShutdownService);
```

Comment:
That's what I meant by `IServiceCollection.Replace`. There is an extension method for that in DependencyInjection.